### DISCUSSION TITLE
Increase endpoint timeout past 60s?

### DISCUSSION THREAD

**Discussion by devchris2303** [postalserver/postal#2866] [Help with using Postal]
Hi there,

First of all I want to say Postal has been such a game changer for what we use it for - so thank you for continuously supporting it!

One question I have is regarding the HTTP Endpoint time out - I see that it maxes out at 60s. To add context, we only use Postal for mail ingress. This works perfectly 99.99999% of the time. However there is the odd case where a user sends in an email with quite a lot of attachments and the email ends up being huge, so much so that the time taken to accept and process is more than 60s. 

Postal naturally gives up and reports a time out, however, our backend system still processes the request and imports the email at , say, 70 seconds - only postal has already given up....then it'll try again and again etc until we have the same email imported 18 times.

I'm thinking of possible solutions here, such as storing the postal mail ID in our backend and if its' imported report a success - maybe a solution. However a quick and dirty solution in the meantime might be to override the timeout somehow?

**Comment by willpower232** [comment]
Can you explain a bit more about how your webhook is set up? Which event are you listening to? I'm not sure how the attachments are involved unless you're actually having difficulties sending the message?

**Reply by devchris2303** [reply]
Ah ever so sorry - I meant HTTP Endpoint! I'll update accordingly.

**Reply by willpower232** [reply]
Ah so you're sending a message with potentially many attachments and then do you eventually get a 502 or something from nginx or caddy?

**Reply by devchris2303** [reply]
The other way - emails coming into Postal and then onto the Endpoint are timing out going to the Endpoint: 

![image](https://github.com/postalserver/postal/assets/38257691/53e29a5b-0f02-4985-9f84-05f1274b9479)

However, our endpoint is importing each file, so it duplicates on every attempt.

**Reply by willpower232** [reply]
Ah right so is it possible for you to not store the attachments on the initial event but fetch them later on? Once you have the Postal message ID, the rest of the API might help?

**Reply by devchris2303** [reply]
Yeah, that could be another solution. 

I've put a fix into our backend that logs the Postal Mail ID when it imports, and then we check for its existence on the first request. 

So if it times out on the first request, but it has gone in - the second attempt will get a 200....where as if it keeps failing the existing functionality will stand.

**Reply by willpower232** [reply]
I believe that the approach of storing the webhook and doing actions on it later is recommended by many software vendors to avoid the issue you're describing.

It looks like the messages API can return attachments if requested

https://apiv1.postalserver.io/structures/message.html
https://apiv1.postalserver.io/controllers/messages/message.html

**Reply by adamcooke** [reply]
As Will says, I'd have the mail processing moved to a background process if it's taking more than a couple of seconds to process.

If you handle this in the foreground of a web process, you'll be tying up a Postal worker and a thread/process on your application's web server for the duration of the processing.

**Reply by devchris2303** [reply]
Roger that, thanks chaps.

I think you're right, we're trying to do too much in one request - which for the large part works fine.

Our live postal system has been up for a couple of weeks and has processed ~7k inbound emails, these are received and processed in seconds. It's just the larger ones that cause trouble, we'll look at modifying this as per your comments. 

On the topic of the API - I have got a PR outstanding to update the docs if you fancy taking a look: https://github.com/postalserver/postal-api/pulls

**Reply by willpower232** [reply]
A great time to ask @adamcooke how the new API is coming along :sweat_smile:

**Reply by adamcooke** [reply]
Soon™️
